crates/oxc_ast/src/ast_builder_impl.rs

```diff
@@ -0,0 +1,219 @@
+    #[inline]
+    pub fn new(allocator: &'a Allocator) -> Self {
+        Self { allocator }
+    }
+
+    #[inline]
+    pub fn alloc<T>(self, value: T) -> Box<'a, T> {
+        Box::new_in(value, self.allocator)
+    }
+
+    #[inline]
+    pub fn new_vec<T>(self) -> Vec<'a, T> {
+        Vec::new_in(self.allocator)
+    }
+
+    #[inline]
+    pub fn new_vec_with_capacity<T>(self, capacity: usize) -> Vec<'a, T> {
+        Vec::with_capacity_in(capacity, self.allocator)
+    }
+
+    #[inline]
+    pub fn new_vec_single<T>(self, value: T) -> Vec<'a, T> {
+        let mut vec = self.new_vec_with_capacity(1);
+        vec.push(value);
+        vec
+    }
+
+    #[inline]
+    pub fn new_vec_from_iter<T, I: IntoIterator<Item = T>>(self, iter: I) -> Vec<'a, T> {
+        Vec::from_iter_in(iter, self.allocator)
+    }
+
+    #[inline]
+    pub fn new_str(self, value: &str) -> &'a str {
+        String::from_str_in(value, self.allocator).into_bump_str()
+    }
+
+    #[inline]
+    pub fn new_atom(self, value: &str) -> Atom<'a> {
+        Atom::from(String::from_str_in(value, self.allocator).into_bump_str())
+    }
+
+    /// # SAFETY
+    /// This method is completely unsound and should not be used.
+    /// We need to remove all uses of it. Please don't add any more!
+    /// <https://github.com/oxc-project/oxc/issues/3483>
+    #[inline]
+    pub fn copy<T>(self, src: &T) -> T {
+        // SAFETY: Not safe (see above)
+        #[allow(unsafe_code)]
+        unsafe {
+            std::mem::transmute_copy(src)
+        }
+    }
+
+    /// Moves the expression out by replacing it with a null expression.
+    #[inline]
+    pub fn move_expression(self, expr: &mut Expression<'a>) -> Expression<'a> {
+        let null_expr = self.expression_null_literal(expr.span());
+        mem::replace(expr, null_expr)
+    }
+
+    #[inline]
+    pub fn move_statement(self, stmt: &mut Statement<'a>) -> Statement<'a> {
+        let empty_stmt = self.empty_statement(stmt.span());
+        mem::replace(stmt, Statement::EmptyStatement(self.alloc(empty_stmt)))
+    }
+
+    #[inline]
+    pub fn move_statement_vec(self, stmts: &mut Vec<'a, Statement<'a>>) -> Vec<'a, Statement<'a>> {
+        mem::replace(stmts, self.new_vec())
+    }
+
+    #[inline]
+    pub fn move_assignment_target(self, target: &mut AssignmentTarget<'a>) -> AssignmentTarget<'a> {
+        let dummy = self.simple_assignment_target_identifier_reference(Span::default(), "");
+        mem::replace(target, dummy.into())
+    }
+
+    #[inline]
+    pub fn move_declaration(self, decl: &mut Declaration<'a>) -> Declaration<'a> {
+        let empty_decl = self.variable_declaration(
+            Span::default(),
+            VariableDeclarationKind::Var,
+            self.new_vec(),
+            false,
+        );
+        let empty_decl = Declaration::VariableDeclaration(self.alloc(empty_decl));
+        mem::replace(decl, empty_decl)
+    }
+
+    /* ---------- Constructors ---------- */
```

Comment:
I guess all code below this will become "templates" or "snippets" in the near future.

Reply:
Yes, these are just convenient special methods that are too useful to get rid of them at this stage.
The memory manipulation methods are here to stay, We want to put more care into writing/documenting them.